### DISCUSSION TITLE
Support for External USB Cams

### DISCUSSION THREAD

**Discussion by bigkukuban** [PhenoApps/Field-Book#629] [Ideas]
In one of our projects, an external camera connected to a smartphone via USB port is used. 
We have tried to use this camera also with Field-Book. 
Unfortunately, at least our camera is not recognised. 
Is there a way to use these cameras from the user's point of view? Or does Field-Book have to be adapted for this?

**Comment by trife** [comment]
It's going to depend on the camera. The library we're currently using has a few compatibility issues that make camera support spotty. I think if the camera has UVC compatibility, it "should" work. What model of camera are you trying to use?

As an aside- we're currently working on adding support for gopro cameras and the early results look positive. In terms of external camera support, we're hoping that this will be sufficient for most breeding programs.

**Reply by Rebeni** [reply]
@trife, regarding your remark on the GoPro support. Would that be directly on the GoPro or a Gopro connected to an android device via USB, bluetooth, wifi?

**Reply by trife** [reply]
We're using the [OpenGoPro](https://gopro.github.io/OpenGoPro/) project to integrate support for the cameras directly into Field Book. We're primarily targeting BT/Wi-Fi, but USB may also be an option.

**Reply by Rebeni** [reply]
Thanks Trife! USB connected would be great, I know it is old-fashioned but reliable in certain ways :)

**Reply by Rebeni** [reply]
@trife , We have been using the USB connected camera in fieldbook. It works nicely BUT we see that the images are all compressed somehow to 1080x810 whereas the cameras used have much higher resolution. Is that something in the code of Fieldbook? Would be nice to understand. Is there a reason for this? Happy to share more details.

**Reply by trife** [reply]
Potentially Field Book issue, potentially the library that we're using. #889 is an open issue to follow progress as we work out the problem

**Comment by Rebeni** [comment]
Hi, I am wondering if this is already possible within Fieldbook?

I have a GoPro and would be able to test that once Fieldbook is able to
connect into an external camera.



Op do 26 okt 2023 om 20:38 schreef Edwin Reidel ***@***.***>:

**Comment by edwinreidel** [comment]
Hi, yes, my understanding is that GoPro is now compatible with Field Book.  Looking forward to hearing how it works out for you, as well as seeing sample images!

Edwin J. Reidel, Ph.D.
Cornell University
Phenomics Coordinator
https://www.breedinginsight.org/

Remotely based in Portland, Oregon

Cell:  +1 (503) 505-8841
Email:  ***@***.******@***.***>

Funded by USDA through Cornell University


From: Rebeni ***@***.***>
Date: Friday, October 27, 2023 at 12:08 AM
To: PhenoApps/Field-Book ***@***.***>
Cc: Edwin J. Reidel ***@***.***>, Manual ***@***.***>
Subject: Re: [PhenoApps/Field-Book] Support for External USB Cams (Discussion #629)
Hi, I am wondering if this is already possible within Fieldbook?

I have a GoPro and would be able to test that once Fieldbook is able to
connect into an external camera.



Op do 26 okt 2023 om 20:38 schreef Edwin Reidel ***@***.***>:

**Comment by chaneylc** [comment]
@edwinreidel @Rebeni @bigkukuban 
Hello all,
Right now this feature is limited to the GoPro Hero11 Black. Feel free to post your specific devices and that will help me triage compatibility for later versions.

That being said, other GoPros might work, but haven't been tested.

**Reply by rlmorgan7** [reply]
Hi, 
I forgot which version was supported, so I just bought the newest GoPro to test the trait ... would it be possible to get the GoPro Hero12 Black supported?

It does seem to start to connect to the 12, but it doesn't complete the action.  

Thanks 
Robyn

**Reply by chaneylc** [reply]
Thanks for testing that Robyn. That is doable, currently I'm working on improving GoPro/USB traits, but it might be a few weeks before I work on integrating new versions of GoPro.

**Comment by Rebeni** [comment]
Dear Edwin,

We have been trying to use the GoPro as well as a USB connected camera
within Fieldbook. Whereas both types of cameras are recognized within
fieldbook and one can take images we encountered issues with each of those,
but different issues.

With a USB cable connected camera we get the images generated but at very
low resolution, 1080x810 most of the times. This seems to depend on the
screen size of the device used and implicates it might take a screenshot
instead of a real image. So even if we use a 13ML USB-camera/webcam or
alike we get roughly 1Mpx images. Too low resolution for our needs.

With the GoPro it actually works well but when taking multiple images
relatively shortly after each other we get the wrong image affiliated to
the wrong plot. In fact sometimes it is all the same image. Seems like the
transfer of images from GoPro to the device is not optimal/too slow.

I was curious if you experience the same thing or not?

Which version of Fielbook do you use?

Thanks for any feedback,
Kind regards
Rene



Op vr 27 okt 2023 om 19:30 schreef Edwin Reidel ***@***.***>:

**Reply by edwinreidel** [reply]
Dear Rene,

The group that I work with hasn't begun using Field Book for image collection.  Hopefully the big bugs can be eliminated before our field season in the Northern Hemisphere.  There is a rumor that the Canon Camera Control API may be under consideration by the developers.  I would greatly prefer to recommend a Canon versus a GoPro for still photography.

I look forward to hearing more from you and from the Field Book team.

My best,
Edwin

**Reply by chaneylc** [reply]
Hello all,
I recently made some improvements to the USB trait, but they have not been released yet. Here is some information:
The USB trait actually only supports UVC (USB Video Class) PNP (Plug and Play) cameras. The library we use to handle this doesn't directly support still capture shots, so the resolution will be based on the camera's video streaming capabilities. I had a similar issue that you had where my resolution was about 1080x800, with my recent improvements you can now select a supported resolution, which I was able to increase to 1920x1080 for my test camera.

I'm also looking at libraries to handle PTP compatible cameras. Currently I have the Canon PowerShotV10 working and I can capture images of 3648x2432. 

My current GoPro implementation using the Hero 11 Black can capture 12.8MB images at 5568x4872

**Reply by Rebeni** [reply]
@chaneylc Would it be possible to use/test the updated version? Or when will it be released?
Thanks!

**Reply by trife** [reply]
@Rebeni The updated version can be built from #898 . This branch will merge a lot of the photo-based phenotyping functions. I can't give an estimate when it will be released.

**Reply by edwinreidel** [reply]
@chaneylc I am able to operate my Canon EOS DSLR with my Android Tablet via Canon's Camera Connect App.  Although my tablet sees the EOS camera's bluetooth signal, I am _unable_ to connect the camera directly to the tablet via Bluetooth ("an app is needed to use this device").

If I make a wired connection between the camera and tablet (HDMI out to USB-C), will it work with Field Book?  Or is the feature limited to PowerShot cameras for now?

Also, does the PowerShotV10 connect wirelessly or by USB only?

**Reply by chaneylc** [reply]
Your Canon EOS should work with Field Book over a wireless connection. Each camera is a little different, so if you run into trouble please let me know. Generally, you should be able to find a Canon menu option that says "Connect to a smartphone (tablet)" then "Add a device..." this will start a WiFi AP. Then, using the Field Book Canon trait, search for this AP and establish a connection/stream.

I don't think you can use Canons as a UVC (wired connection) unless you use a third-party streaming chip/cable which would just broadcast the screen of your camera.

**Reply by edwinreidel** [reply]
Thanks, I have successfully connected my Canon EOS via WiFI and Bluetooth connections using Camera Connect.  I can operate the camera wirelessly with my iPhone and with an Android tablet via the Camera Connect App.  

Although the camera is listed as an available Bluetooth device in the menu on my Galaxy Tab running Field Book, I cannot however connect the camera directly to the tablet (without an app).  When attempting to establish a Bluetooth connection, I get the following message:  "An app is needed to use this device" [via bluetooth].

I will try to add a video or screenshots if that helps.

**Reply by trife** [reply]
@edwinreidel The Canon code isn't going to be live until v6. We're wrapping up a few other features for this but I'm hoping it's available for testing by the end of the month.
